File: skills.py
from __future__ import annotations

from pathlib import Path


class Skills:
    def __init__(self, root: str = "skills"):
        self.root = Path(root)
        self.loaded: dict[str, str] = {}

    def list(self) -> list[dict[str, str]]:
        items = []
        for skill_md in sorted(self.root.glob("*/SKILL.md")):
            name = skill_md.parent.name
            description = self._description(skill_md)
            items.append({"name": name, "description": description})
        return items

    def catalog_text(self) -> str:
        lines = ["Available skills:"]
        for item in self.list():
            lines.append(f"- {item['name']}: {item['description']}")
        lines.append("Load a skill with skill_load before following it.")
        return "\n".join(lines)

    def load(self, name: str) -> str:
        path = self.root / name / "SKILL.md"
        if not path.exists():
            raise FileNotFoundError(f"Unknown skill: {name}")
        body = path.read_text(encoding="utf-8")
        self.loaded[name] = body
        return body

    def _description(self, path: Path) -> str:
        for line in path.read_text(encoding="utf-8").splitlines():
            if line.lower().startswith("description:"):
                return line.split(":", 1)[1].strip().strip('"')
        return "No description."

File: src/skills.rs
use anyhow::{bail, Result};
use std::fs;
use std::path::PathBuf;

#[derive(Clone, Debug)]
pub struct Skills {
    root: PathBuf,
}

impl Skills {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }

    pub fn catalog_text(&self) -> String {
        let mut lines = vec!["Available skills:".to_string()];
        if let Ok(entries) = fs::read_dir(&self.root) {
            for e in entries.flatten() {
                let skill = e.path().join("SKILL.md");
                if skill.exists() {
                    let name = e.file_name().to_string_lossy().into_owned();
                    let desc = fs::read_to_string(&skill)
                        .ok()
                        .and_then(|t| {
                            t.lines()
                                .find(|l| l.to_lowercase().starts_with("description:"))
                                .map(|l| l.splitn(2, ':').nth(1).unwrap_or("").trim().to_string())
                        })
                        .unwrap_or_else(|| "No description.".into());
                    lines.push(format!("- {name}: {desc}"));
                }
            }
        }
        if lines.len() == 1 {
            lines.push("- none yet (self-learning will create them)".into());
        }
        lines.push("Load a skill with skill_load before following it.".into());
        lines.join("\n")
    }

    pub fn load(&self, name: &str) -> Result<String> {
        let path = self.root.join(name).join("SKILL.md");
        if !path.exists() {
            bail!("Unknown skill: {name}");
        }
        Ok(fs::read_to_string(path)?)
    }

    pub fn write_skill(&self, name: &str, body: &str) -> Result<()> {
        let dir = self.root.join(name);
        fs::create_dir_all(&dir)?;
        fs::write(dir.join("SKILL.md"), body)?;
        Ok(())
    }
}
